### DISCUSSION TITLE
Dynamically  update Data between  tabs

### DISCUSSION THREAD

**Discussion by bajirama** [zauberzeug/nicegui#3414] [Q&A]
### Question

I'm in need of assistance with this issue and find myself going down a rabbit hole, unable to find a solution. Typically, with NiceGUI, a few hours of work yields some resolution, but I've been struggling for days now.

 Here's the scenario: I have two tabs on the left. In the first tab, I enter a value that is displayed in the AGGrid. The key value will be the screen name, which remains constant. The '**Add'** button will append the value to the grid, while the **'Save'** button will store the value in app.storage.user.

 In the second tab, I need to reference the same data entered in the first tab. Any changes in Tab 1 should be reflected in Tab 2. Additionally, I need the value from Tab 1 to be reflected in other tabs on the left panel.

 For clearer code, I plan to separate each tab into its own module, like Tab 2. Currently, I'm using app.storage.user to save values in the .nicegui folder. The values are saved, and changes are updated as needed in app.storage.user. 

However, my challenge is dynamically refreshing and retrieving the latest values from app.storage.user. 
Once resolved, I also plan to add a dropdown at the top section of the left panel, allowing selection of any value to reset all tabs for new criteria. 

Could you advise on how to achieve dynamic refresh for tabs? I've attached screenshots, minimal viable code, and sample values from app.storage.user for reference.

**First time entry  
Tab 1  - Add Entry**

![image](https://github.com/user-attachments/assets/5293dcaa-eec2-462a-ba2e-a2af52c4b6ee)

**Tab2 – Update Entry** 
![image](https://github.com/user-attachments/assets/269abded-e053-4b5f-b8fe-a3d87efaa298)

**After Screen Refresh**  
**Tab 1  - Add Entry**
![image](https://github.com/user-attachments/assets/72606c52-1ccf-461e-a804-c1a1b143728a)

**Tab2 – Update Entry** 
![image](https://github.com/user-attachments/assets/d2e0905f-2a31-41cd-a808-de544ce6b57c)

**Sample  app.storage.user** 
```json
{"MyKeyName": [
        {
            "acc_type": "Value11",
            "acc_no": "Value12",
            "AccName": "Value13",
            "region": "Value14"
        },
        {
            "acc_type": "Value21",
            "acc_no": "Value22",
            "AccName": "Value23",
            "region": "Value24"
        },
        {
            "acc_type": "Value31",
            "acc_no": "Value32",
            "AccName": "Value33",
            "region": "Value34"
        },
        {
            "acc_type": "Value41",
            "acc_no": "Value42",
            "AccName": "Value43",
            "region": "Value44"
        }
    ]
}
```
**Minimum Viable Code** 
**refreshable.py**
```py
from nicegui import ui, app ,Client
import _tabfun

@ui.page('/')
def page(client: Client):
    accdetails = []
    @ui.refreshable
    def refreshStoage() -> None:
        accdetails = app.storage.user.get("MyKeyName")
        print(f"refresh {accdetails}")
        ui.label("Upated")

    def savestorage():
        app.storage.user[new_key.value] = accdetails
        refreshStoage.refresh()

    with ui.left_drawer().style('background-color: #F8F8F8').props("elevated v-model='leftDrawerOpen'"):

        with ui.tabs().props('vertical').classes('w-full') as tabs:
            tab1 = ui.tab('Add Entry')
            tab2 = ui.tab('Update Entries')
    with client.content:
      with ui.tab_panels(tabs, value=tab1):
          with ui.tab_panel(tab1):

              def add():
                  laccdetails = [{'acc_type': new_value1.value,
                                 'acc_no': new_value2.value,
                                 'AccName': new_value3.value,
                                 'region': new_value4.value,
                                 }]
                  accdetails.extend(laccdetails )
                  AccDetTable.options['rowData'] = accdetails
                  AccDetTable.update()


              with ui.card().style('width: 1400px; max-width: none'):
                 with ui.row().classes("w-full"):
                     accDetColDef = [
                            {'headerName': "Acc Type ", "field": "acc_type", 'checkboxSelection': True},
                            {'headerName': "Account Number", "field": "acc_no"},
                            {'headerName': "Account Name", "field": "AccName"},
                            {'headerName': "Region", "field": "region"},
                        ]
                     AccDetTable = ui.aggrid({
                            'defaultColDef': {'flex': 1},
                            'columnDefs': accDetColDef,
                            'rowData': accdetails,
                           'rowSelection': 'single',
                         }).classes('max-h-40')

                 with ui.row().classes("w-full"):
                    new_key    = ui.input("KeyValue", value= "MyKeyName",).props("disable")
                    new_value1 = ui.input('Value1')
                    new_value2 = ui.input('Value2')
                    new_value3 = ui.input('Value3')
                    new_value4 = ui.input('Value4')
                    refreshStoage()
                    ui.button('Add', on_click=add)
                    ui.button("Save", on_click=savestorage)

          with ui.tab_panel(tab2):
            _tabfun.tab2scr()


ui.run(storage_secret='MY_PORTAL')
```
**Tab 2  - Module  - _tabfun.py**

```py
from nicegui import ui, app

def tab2scr():
    laccDet = app.storage.user.get("MyKeyName")
    with ui.row().classes('w-full'):
      with ui.card().style('width: 1400px; max-width: none'):
         ui.notify("Test")
         ui.label(laccDet)
         accDetColDef = [
                {'headerName': "Acc Type ", "field": "acc_type", 'checkboxSelection': True},
                {'headerName': "Account Number", "field": "acc_no"},
                {'headerName': "Account Name", "field": "AccName"},
                {'headerName': "Region", "field": "region"},
              ]
         AccDetTable = ui.aggrid({
           'defaultColDef': {'flex': 1},
           'columnDefs': accDetColDef,
           'rowData': laccDet,
           'rowSelection': 'single',
        }).classes('max-h-40')


```

**Comment by rodja** [comment]
Sorry @bajirama. This code is still not minimal. Can you boil it down further? No drawer, lo layout, not multiple values, etc. just the bare minimum of what you need to show the issue.

**Comment by bajirama** [comment]
I understand the code is a bit long, but my problem is building a complex UI screen where I have tab options on the left and content on the right and want to share the data between tab planes dynamically. 

For code clarity and maintainability, modularize each tab panel into different modules.

Can you please let me know how I can share data dynamically between tabs? I am really stuck

**Comment by rodja** [comment]
Sorry, we generally do not have the time to debug such applications. But have you seen our [chat example](https://github.com/zauberzeug/nicegui/blob/main/examples/chat_app/main.py)? It is a small demo where multiple tabs can be opened and everyone sees the same content.

**Comment by Student1098283** [comment]
Hello @bajirama , did you finally find a solution to this problem? I have the same issue and I'm kind of stuck on what to do.

**Comment by evnchn** [comment]
So you want to share the code between the two tabs, so that you can (a) adhere to DRY principle and (b) share the data. 

How about I counter-propose that you can also implement the "Upsert" operation, which means to "[To insert rows into a database table if they do not already exist, or update them if they do](https://en.wiktionary.org/wiki/upsert)". Well it isn't a database but `app.storage.user`, but the same spirit applies. With NiceGUI, because you can update the button text, you can make the "Add" button change into "Update details for John" the moment John's `acc_no` is filled in. 

While we are at it, do you notice that `app.storage.user.get("MyKeyName")` is a list? If you want to find, say, John's account, it will take O(n) time. Not good. You know how much I hate O(n) operation for a single item (Think: https://github.com/zauberzeug/nicegui/pull/4463). If you agree with me and consider that `acc_no` can be considered a key field, then make `app.storage.user.get("MyKeyName")` a dictionary key with `acc_no` and value being a dict of `acc_type`, `AccName`, and `region`

But the issue of, User A updating the information, User B does not see it until refresh... You can try: 
* Periodically call [`update_rows`](https://nicegui.io/documentaion/table#table) on the table using [`ui.timer`](https://nicegui.io/documentation/timer#timer) 
* Let the `ui.table` inside [`ui.refreshable`](https://nicegui.io/documentation/refreshable) and run `.refresh()` to update the table
* Multicast to the other clients of the page via [`for client in app.clients('/'):`](https://nicegui.io/documentation/page#multicasting) and mess about, potentially to [`ui.notify("Your page is outdated, refresh to see latest data from User A")`](https://nicegui.io/documentation/notify) or even [`ui.navigation.reload()`](https://github.com/zauberzeug/nicegui/blob/1347b6956b0cc245464f8b44e94d88cc9c0c9913/nicegui/functions/navigate.py#L33)

There's still much more to discover for NiceGUI, and I am glad that you have a need, and we have many solutions. 

And no. It doesn't seem like that there is a bug or deficiency with NiceGUI, so far from what I am seeing. 

@falkoschindler `ui.navigate.reload` not in https://nicegui.io/documentation/navigate#navigation_functions?

_PS: I can already imagine that you need to delete rows. Not to fret, just add a delete button that only looks at the `acc_no` and delete the row. No need to add another tab._

**Reply by falkoschindler** [reply]
> `ui.navigate.reload` not in https://nicegui.io/documentation/navigate#navigation_functions?

Fixed it in bb5edf45424846f76ffee4406931a6f97b349a80.
